### DISCUSSION TITLE
Zones overhaul

### DISCUSSION THREAD

**Discussion by joshuavictorchen** [terrapower/armi#811] [General]
The `armi.reactor.zones` module is very SAS4A/SASSYS-1-specific and is in need of an overhaul - see [this issue](https://github.com/terrapower/armi/issues/563).

This discussion serves to communicate these requirements and manage progress/collaboration on implementation.

**Note**:  In the following bullets we define the term `Item` to be either an `Assembly` or a `Block`. And probably, one `Zone` collection will have to be defined as _all_ `Assembly` objects or _all_ `Block` objects, with no mixing (just to make things easier).

## Requirements

The following requirements describe what a good `zones` module should look like. 

* [ ]  `zones` should be an attribute of the `Core` that tracks `Item` locations AND the names of the `Items` in those locations
  * [ ] This means that `Zones` will be affected if any `Item`s are moved, but the `Item`s themselves can be perturbed without disturbing the `Zones` (i.e., updating their `blocks`’ reactivity coefficient parameters)
  * [ ] The primary function of `zones` will be to contain a list/set/collection of `Zone` objects.
* [ ] A `Zone` should have the following attributes:
  * [ ] `name` (string)
  * [ ] `locations` (dict of locations and corresponding `Item` names)
  * [ ] `length` (int of count of locations)
  * [ ] `id` (hash based on the `locations` attribute), or a `hash()` function
* [ ] `Zones` may be mutable, but their `id` attributes should be updated if any changes are made to them, in order to:
  * [ ] Validate that reactivity coefficient parameters were calculated for the same location-`Item` configuration as expected by the SAS model
  * [ ] Adjust `Zones` on the fly if needed (i.e., for sensitivity studies)
* [ ] `Zones` may have the following utility functions:
  * [ ] `addItem(item, toZone)`
  * [ ] `transferItem(item, fromZone, toZone)`
  * [ ] `removeItem(item, fromZone)`
  * [ ] `checkDuplicates()`
  * [ ] `zoningStrategy()` <-- abstract method?
* [ ] Users should be able to easily get specific `Item` objects described by the `Zone`
* [ ] `Zones` should be generated whenever a reactor database is created or loaded
  * [ ] `Zones` should be able to be written to/read from the database
* [ ] `Zones` should NOT be able to contain duplicate `Items`
* [ ] `Zones` should NOT need to be contiguous or contain contiguous `Items`
* [ ] `Zones` should NOT need to be comprehensive (i.e., `zones` don't have to contain ALL of the `Items` in the `Core`)

## Implementation details:

* Unit tests should be written to ensure `Zones`
  * [ ] are easy to implement as a `Plugin` or `Interface` or something easy
  * [ ] can be saved to a DB
  * [ ] can be read from a DB

**Comment by joshuavictorchen** [comment]
Thanks to @john-science for helping to refine the above requirements.

The feature branch for this overhaul is: https://github.com/terrapower/armi/tree/zones_overhaul

**Comment by john-science** [comment]
Awesome!  

Let's try and organize our work a bit.  Just edit this comment and put your name next to the thing you're working on:

* [x] Removing various current zoning strategies (**JOHN**)
  * [X] Removing all references to Hot Zones (**JOHN**)
  * [x] Removing all references to Ring Zones (**JOHN**)
* [x] First try at flushing out the `Zone` and `Zones` classes. (**JOHN**)
* [X] Changing how the zoningStrategy Setting works (**JOHN**)
  * [X] Move the zoning strat to a Plugin hook.
  * [ ] Add Zoning Strategy example to the docs (**JOHN**)
* [X] Refactor Zone Summary  (**JOHN**)
  * [X] Should we just print something simpler that you could copy/paste into manual zone definitions? (**JOHN**)
  * [X] Remove duplicate Zone summary at: `armi/bookkeeping/report/reportingUtils.py`? (**JOHN**)
  * [X] This thing ONLY works for Axial Reactors, and also not `Blocks`. (**JOHN**)
* [X] ~Do we need to remove anything from the ARMI Docs? (**NO**)~
* [x] Implement zoning strategy for SAS/NATD use cases within the SAS Interface (**JOSH**)
* [X] Make multiple Zoning Strategies `raise` and `Error`, rather than just a `warning`? (**JOHN**)
* [X] Rename `defineZoningStrategy()` to `applyZoningStrategy()`? (**JOHN**)
* [ ] TBD...

**Reply by john-science** [reply]
@joshuavictorchen  Okay, I have torn out a bunch of code. Mostly, I tore out all those old, strange "zoning strategies" that were applied by default. And I tore out some odd references to "Rings" that existed by default in the Zone class.  (Not all Reactors might have rings, after all.)

So, we have a lot less cruft lying around in the `zones.py` class.  But, of course, at some point, this will mean one or two downstream plugins will have to define their own Zoning Strats.  But that's part of the Requirement.

**Reply by joshuavictorchen** [reply]
I'm thinking of kind of a hybrid scenario re: changing how the zoningStrategy Setting works - would it be possible to let the user select a specific plugin to implement their zoning strategy via a Setting?

Then that plugin's implementation of an abstract `applyZoningStrategy` class (or similar) would be called to create the `Zones`.

In my mind, this would also prevent plugins from fighting with/overriding each other's zones.

**Reply by john-science** [reply]
Okay, I was wondering how we were going to make the zoning strategies work.  I had some vague ideas, but yours is better now that I look at it.

So, let me take a crack at this tonight/tomorrow morning. I looks like we would put a new `abstractmethod` here:
 
https://github.com/terrapower/armi/blob/9a510ca3a330c04471fa49f1cf7bf48d5f646f2b/armi/plugins.py#L141

And add another `abstractmethod` / `HOOK` thing, like one of these:

https://github.com/terrapower/armi/blob/9a510ca3a330c04471fa49f1cf7bf48d5f646f2b/armi/plugins.py#L345-L349

https://github.com/terrapower/armi/blob/9a510ca3a330c04471fa49f1cf7bf48d5f646f2b/armi/physics/fuelCycle/__init__.py#L77-L80

And maybe call our new hook in `buildZones()`:

https://github.com/terrapower/armi/blob/9a510ca3a330c04471fa49f1cf7bf48d5f646f2b/armi/reactor/zones.py#L392

In the same way the other hooks get called:

https://github.com/terrapower/armi/blob/9a510ca3a330c04471fa49f1cf7bf48d5f646f2b/armi/apps.py#L139

And maybe we build a nice example for people, and put it in the docs. So everyone can see how to define their own zoning strategy in the future.

**Comment by john-science** [comment]
### Minor Requirement Change

After looking at the code, I think we can simplify our design significantly if we allow a `Zone` to be a collection of either `Assemblies` or `Block`s.  The biggest reason for this is that both `Assembly` and `Block` have a `getLocation()` method.  So, we could use those.  The only difference we have to add one piece of data to the system: Each `Zone` would need to say with it is Assemblies or Blocks. 

This would include where we read in Zone definitions. So, this isn't as simple as the old "Assembly-only" pathway, but it is worlds simpler than our original idea of allowing Component-level Zones.

**Reply by joshuavictorchen** [reply]
From my (selfish) Safety perspective, we just need the granularity to get down to the `Assembly` level - so I am 100% for this.

**Comment by john-science** [comment]
Also, @joshuavictorchen , I note that we already have logic in the code base for manually defining `Zones` in the Settings file:

https://github.com/terrapower/armi/blob/e336badf27f2a2c448dc2ac935b065a0c5134b0f/armi/settings/fwSettings/globalSettings.py#L139-L149

And we even check for manual zone definitions:

https://github.com/terrapower/armi/blob/e336badf27f2a2c448dc2ac935b065a0c5134b0f/armi/operators/settingsValidation.py#L336-L345

And they are really simple `["zone1:loc1,loc2,loc3", "zone2:loc5,loc6"]`:

https://github.com/terrapower/armi/blob/e336badf27f2a2c448dc2ac935b065a0c5134b0f/armi/reactor/zones.py#L482-L491

I _think_ we can re-use nearly all of that.  The only thing we need to add in is some way to say if the `Zone` is a group of `Assembly` or `Block` objects.

We will need to change the default to `None` or maybe `"manual"` or something.
